### PR TITLE
Allow ?! as a non-capturing negative lookahead in TreeRegexp

### LINKED ISSUE #481: cucumber-expressions: Cucumber treats a (?!text) group as a capturing group; IndexError is then thrown
## Summary
Given a step definition with a regexp that contains a negative look ahead and a capturing group back to back, causes cucumber to fail to load and gives an unhelpful error message "index 2 out of matches (IndexError)". This is the step definition:

```ruby
And(/^the Projected Answer Time shown in the response accounts for the timezone (?!offset)(.*)$/) do |timezone|
....
end
```

## Expected Behavior

This step definition worked with v2.4.0, it should still work or provide a more detailed error message that would help resolve the problem.

## Current Behavior

An index 2 out of matches (IndexError) error is raised:

~~~
index 2 out of matches (IndexError)
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/group_builder.rb:20:in `offset'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/group_builder.rb:20:in `build'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/group_builder.rb:19:in `block in build'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/group_builder.rb:19:in `map'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/group_builder.rb:19:in `build'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/tree_regexp.rb:51:in `match'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/argument.rb:10:in `build'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-expressions-5.0.17/lib/cucumber/cucumber_expressions/regular_expression.rb:32:in `match'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/glue/step_definition.rb:101:in `arguments_from'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/glue/registry_and_more.rb:66:in `block in step_matches'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/glue/registry_and_more.rb:65:in `each'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/glue/registry_and_more.rb:65:in `reduce'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/glue/registry_and_more.rb:65:in `step_matches'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/step_match_search.rb:19:in `call'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/step_match_search.rb:19:in `call'
/Users/madkins/.rvm/rubies/ruby-2.4.2/lib/ruby/2.4.0/delegate.rb:83:in `method_missing'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/step_match_search.rb:62:in `call'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:66:in `matches'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:47:in `result'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:37:in `find_match'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:33:in `attempt_to_activate'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:29:in `map'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:29:in `new_test_steps'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:23:in `test_case'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/filters/activate_steps.rb:12:in `test_case'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/test/case.rb:25:in `describe_to'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/test/filters/locations_filter.rb:18:in `block in done'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/test/filters/locations_filter.rb:17:in `each'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/test/filters/locations_filter.rb:17:in `done'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/filter.rb:62:in `done'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/test/filters/tag_filter.rb:18:in `done'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/compiler.rb:24:in `done'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core/gherkin/parser.rb:37:in `done'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core.rb:32:in `parse'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-core-3.1.0/lib/cucumber/core.rb:21:in `compile'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/runtime.rb:74:in `run!'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/lib/cucumber/cli/main.rb:33:in `execute!'
/Users/madkins/.rvm/gems/ruby-2.4.2/gems/cucumber-3.1.0/bin/cucumber:9:in `<top (required)>'
/Users/madkins/.rvm/gems/ruby-2.4.2/bin/cucumber:23:in `load'
/Users/madkins/.rvm/gems/ruby-2.4.2/bin/cucumber:23:in `<main>'
~~~

## Steps to Reproduce (for bugs)

1. Define the error raising step:
```ruby
And(/^foo is a (?!big)(.*)$/) do |bar|
  puts bar
end
```
2. Define an ambiguous step to show the need for this to work:
```ruby
And(/^foo is a big (.*)$/) do |bar|
  puts bar
end
```
3. Define 2 scenarios that utilize the steps:
```gherkin
  Scenario: Cucumber fails to load
    Given foo is a bar
    
  Scenario: Used to prevent ambiguous steps
    Given foo is a big bar
```
4. Run the tests: `bundle exec cucumber`

## Your Environment

* Version used: 3.10
* Operating System and version: MacOS High Sierra 10.13.3

## Related Issue: https://github.com/cucumber/cucumber/issues/329

### PR DESCRIPTION
**Pull request by luke-hill**
This is considered to be a regression from cucumber 2.4.

Attempts to close #481 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by aslakhellesoy** [comment]
Thanks for looking into this @luke-hill! We need to update the other implementations before this can be merged.

**Comment by luke-hill** [comment]
This is already fixed in java I believe as @mpkorstanje pointed out in the original issue. I'm not sure on other languages.

**Comment by mpkorstanje** [comment]
I don't recall fixing this.

**Comment by aslakhellesoy** [comment]
As long as we maintain multiple implementations of the same library, master should always contain equivalent implementations/tests.

Evolving them independently should be avoided at all costs.

**Comment by luke-hill** [comment]
@mpkorstanje when speaking to me at Cukenfest, you linked me to the relevant file which already had `?!` marked as non-capturing in Java. So it doesn't need fixing there as it isn't broken there.

No idea on the other languages though

**Comment by mpkorstanje** [comment]
Aah. Right. Should sync the tests too.

**Comment by luke-hill** [comment]
Agreed, I had to change the existing tests (Well I kind of didn't, kind of did), just so I understood what was happening and could make them pass.

The two tests that should pass in all languages are
`/a(?:b)(c)/ - matches to abc and has 1 capture c`
`/a(?!b)(.+)/ - matches to aBc and has 1 capture c`

cc/ @jaysonesmith can you look into go and then we need to check javascript and what else?

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/tree_regexp.rb [hunk 1]
```diff
@@ -15,36 +15,36 @@ def initialize(regexp)
         non_capturing_maybe = false
         char_class = false
 
-        @regexp.source.split('').each_with_index do |c, n|
-          if c == '[' && !escaping
```
Comment (aslakhellesoy) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/tree_regexp.rb]: Please don't make changes (such as renames) unless they are related to fixing the bug. Smaller diffs makes it much easier to review and port to other languages.

Comment (luke-hill) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/tree_regexp.rb]: Fixed this up, left the top bit as a slight optimisation

**Comment by luke-hill** [comment]
I've standardised the tests in JS/Java/GO so they all test identical items. Could you give it a quick look over (I'm fairly confident in the Java / JS ones, less so on the GO one).

cc/ @jaysonesmith / @mpkorstanje

**Comment by jaysonesmith** [comment]
If they pass the expected assertions, looks good to me @luke-hill! Thanks :)

**Comment by jaysonesmith** [comment]
Just fix your conflict :p

File: cucumber-expressions/go/tree_regexp_test.go [hunk 2]
```diff
@@ -25,6 +25,20 @@ func TestTreeRegexp(t *testing.T) {
...
 	})
+	
+  t.Run("ignores `?:` as a non-capturing group", func(t *testing.T) {
+    tr := NewTreeRegexp(regexp.MustCompile("a(?:b)(c)"))
+    group := tr.Match("abc")
+    require.Equal(t, *group.Value(), "abc")
+    require.Len(t, group.Children(), 1)
+  })
+
+  t.Run("ignores `?!` as a non-capturing group", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jaysonesmith) [reply cucumber-expressions/go/tree_regexp_test.go]: Ah, good call. @luke-hill Can you see if your IDE/Editor has Go formatting? If you're using VS Code the Go extension is awesome. It'll ask you to install the go tools which you should do and then things like formatting will be done on save.

[the remark this answers is not in the document]
Comment (luke-hill) [reply cucumber-expressions/go/tree_regexp_test.go]: I used RubyMine :P I'll have a look into it. I think there will be a go extension (There is a js / java one)

**Comment by luke-hill** [comment]
p.s. I only standardised the tests, I didn't check the physical code-base. So the fact the builds are failing may be indicative of me needing to fix the tests up. I will check these today.

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java [hunk 3]
```diff
@@ -53,7 +53,7 @@
                 nonCapturingMaybe = false;
             } else if (c == '?' && last == '(') {
                 nonCapturingMaybe = true;
-            } else if (c == ':' && nonCapturingMaybe) {
+            } else if ((c == ':' || c == '!') && nonCapturingMaybe) {
```
Comment (benariss) [inline cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java]: This looks to have fixed this for negatives but not positive lookaheads ('(?=...)'), or look behinds ('(?<=..)' and '(?<!...)').

Comment (benariss) [inline cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java]: This looks to have fixed this for negatives but not positive lookaheads ('(?=...)'), or look behinds ('(?<=..)' and '(?<!...)').

Comment (luke-hill) [reply cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/TreeRegexp.java]: @benariss - Can you create a separate standalone issue for this, I will tag it accordingly.

**Comment by jaysonesmith** [comment]
Looks like my change needs to be tweaked a little. Will investigate.

**Comment by luke-hill** [comment]
Right. Thanks a lot @jaysonesmith for that. This is now RtM.

I'll leave it a day before merging incase there's any last minute requests from anyone monitoring the thread. But once this is in it represents a decent fix across all cucumber languages.

**Comment by jaysonesmith** [comment]
Huzzah! You're welcome and thank you for all your work!

**Pull request merged**